File: core/java/src/main/java/io/github/cyfko/dynamicfilter/core/validation/PropertyRef.java
package io.github.cyfko.dynamicfilter.core.validation;

import java.util.Set;

/**
 * Interface for property references in dynamic filtering.
 * 
 * Developers should create their own enums implementing this interface to define
 * the properties available for their entities.
 * 
 * PropertyRef contains only the logical property definition (type, operators).
 * Each adapter is responsible for interpreting PropertyRef and creating
 * appropriate conditions. This allows maximum flexibility - a single PropertyRef
 * can map to multiple entity fields or complex conditions.
 * 
 * Example usage:
 * <pre>
 * public enum UserPropertyRef implements PropertyRef {
 *     USER_NAME(String.class, Set.of(LIKE, EQ, IN)),
 *     USER_AGE(Integer.class, Set.of(EQ, GT, GTE, LT, LTE, BETWEEN)),
 *     USER_STATUS(String.class, Set.of(EQ, NE, IN));
 *     
 *     private final Class<?> type;
 *     private final Set<Operator> supportedOperators;
 *     
 *     UserPropertyRef(Class<?> type, Set<Operator> supportedOperators) {
 *         this.type = type;
 *         this.supportedOperators = Set.copyOf(supportedOperators);
 *     }
 *     
 *     @Override
 *     public Class<?> getType() { return type; }
 *     
 *     @Override
 *     public Set<Operator> getSupportedOperators() { return supportedOperators; }
 * }
 * </pre>
 */
public interface PropertyRef {
    
    /**
     * Gets the type of this property.
     * 
     * @return The property type
     */
    Class<?> getType();
    
    /**
     * Gets the set of operators supported by this property.
     * The returned set should be immutable.
     * 
     * @return An immutable set of supported operators
     */
    Set<Operator> getSupportedOperators();
    
    /**
     * Checks if this property supports the given operator.
     * 
     * @param operator The operator to check
     * @return true if the operator is supported, false otherwise
     */
    default boolean supportsOperator(Operator operator) {
        return getSupportedOperators().contains(operator);
    }
    
    /**
     * Validates that the given operator is supported by this property.
     * 
     * @param operator The operator to validate
     * @throws IllegalArgumentException if the operator is not supported
     * @throws NullPointerException if the operator is null
     */
    default void validateOperator(Operator operator) {
        if (operator == null) {
            throw new NullPointerException("Operator cannot be null");
        }
        
        if (!supportsOperator(operator)) {
            throw new IllegalArgumentException(
                String.format("Operator %s is not supported for property %s. Supported operators: %s",
                            operator, this, getSupportedOperators()));
        }
    }
}

File: core/csharp/Validation/PropertyRef.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DynamicFilter.Core.Validation
{
    /// <summary>
    /// Interface for property references in dynamic filtering.
    /// 
    /// Developers should create their own enums implementing this interface to define
    /// the properties available for their entities.
    /// 
    /// Example usage:
    /// <code>
    /// public enum UserPropertyRef : IPropertyRef
    /// {
    ///     UserName,
    ///     UserAge,
    ///     UserEmail
    /// }
    /// 
    /// // PropertyRef implementation
    /// public class UserPropertyRefImpl
    /// {
    ///     public static readonly IPropertyRef UserName = new PropertyRefImpl("string", new[] { Operator.Like, Operator.Equals });
    ///     public static readonly IPropertyRef UserAge = new PropertyRefImpl("int", new[] { Operator.Equals, Operator.GreaterThan });
    ///     public static readonly IPropertyRef UserEmail = new PropertyRefImpl("string", new[] { Operator.Equals, Operator.Like });
    /// }
    /// </code>
    /// </summary>
    public interface IPropertyRef
    {
        /// <summary>
        /// Gets the type of this property.
        /// </summary>
        string Type { get; }

        /// <summary>
        /// Gets the set of operators supported by this property.
        /// </summary>
        IReadOnlyList<Operator> SupportedOperators { get; }

        /// <summary>
        /// Checks if this property supports the given operator.
        /// </summary>
        /// <param name="operator">The operator to check</param>
        /// <returns>True if the operator is supported, false otherwise</returns>
        bool SupportsOperator(Operator @operator);

        /// <summary>
        /// Validates that the given operator is supported by this property.
        /// </summary>
        /// <param name="operator">The operator to validate</param>
        /// <exception cref="ArgumentException">Thrown if the operator is not supported</exception>
        void ValidateOperator(Operator @operator);
    }

    /// <summary>
    /// Implementation of IPropertyRef interface for enum values.
    /// </summary>
    public class PropertyRefImpl : IPropertyRef
    {
        public string Type { get; }
        public IReadOnlyList<Operator> SupportedOperators { get; }

        public PropertyRefImpl(string type, Operator[] supportedOperators)
        {
            Type = type ?? throw new ArgumentNullException(nameof(type));
            SupportedOperators = supportedOperators?.ToList().AsReadOnly() ?? throw new ArgumentNullException(nameof(supportedOperators));
        }

        public bool SupportsOperator(Operator @operator)
        {
            return SupportedOperators.Contains(@operator);
        }

        public void ValidateOperator(Operator @operator)
        {
            if (!SupportsOperator(@operator))
            {
                throw new ArgumentException(
                    $"Operator '{@operator}' is not supported for this property. " +
                    $"Supported operators: {string.Join(", ", SupportedOperators)}");
            }
        }

        public string GetDescription()
        {
            return $"PropertyRef({Type})";
        }

        public override string ToString()
        {
            return $"PropertyRef{{Type={Type}, SupportedOperators=[{string.Join(", ", SupportedOperators)}]}}";
        }

        public override bool Equals(object obj)
        {
            if (obj is PropertyRefImpl other)
            {
                return Type == other.Type;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Type);
        }
    }
}
